### PR TITLE
Add Employee Rostering Backend blog post

### PR DESCRIPTION
**Pull request by cuijulian**

### CODE REVIEW

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+:page-interpolate: true
+:awestruct-author: jucui
+:awestruct-layout: blogPostBase
+:awestruct-tags: [employee rostering, coding, algorithm, production]
+
+Recently, the OptaPlanner team is shifting towards cloud-native applications.
+As a part of this effort, we migrated the OptaWeb Employee Rostering Backend Application from WildFly to Spring Boot.
+
+OptaWeb Employee Rostering is a web application that optimizes shift scheduling for maximum efficiency and fairness for employees.
+In this post we’ll go over how the OptaPlanner web application was moved to the Spring framework, and the key features that were added on top of the old backend.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: I'd prefer "existing" to "old". :-)

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 2]
```diff
@@ -0,0 +1,195 @@
+= Migrating OptaWeb Employee Rostering Backend Application to Spring Boot
+:page-interpolate: true
+:awestruct-author: jucui
+:awestruct-layout: blogPostBase
+:awestruct-tags: [employee rostering, coding, algorithm, production]
+
+Recently, the OptaPlanner team is shifting towards cloud-native applications.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: I'm not a native speaker, but I believe the correct tense after `recently` has to be `has been shifting`.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+== Domain Model
+
+Before we dive into the code, we first need a domain model.
+A domain model represents the entities (Java objects) and how they relate with other entities in the application.
+It tells us the behaviour of the objects and how they change during solving.
+Here's a class diagram of the Employee Rostering model, where employees are nurses:
+
+image::employeeRosteringClassDiagram.png[]
+
+Each of the blocks above represents a Java entity object in our application, which will be saved in a Spring JPA Database for solving.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: I'd rephrase. `which will be saved in a database for solving. We use Spring JPA as our persistence layer.`

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+The arrows show how certain entities map to others, so changing one during the solving process may affect another.
+The `*` and `1` symbols on the arrows represent a many-to-one mapping of the entities in the database.
+
+Here's a simple example of the `Skill` entity:
+
+[source,java]
+----
+@Entity
+@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
+public class Skill extends AbstractPersistable {
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: I'd consider if we want to expose the internals, such as `AbstractPersistable`. Without explaining what it is, it may be confusing to a first time reader. And if you explain it, the blog will be too detailed. 

If we call this a simplified entity, perhaps we can get away with removing these details?

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+We see that each Tenant contains the `id`, `name`, and `version` fields, which are saved in the database whenever a Tenant is added.
+Also note the `Try it out` button in the top right corner.
+We can use it to send HTTP requests with ease for testing purposes.
+Let's try it out:
+
+image::executeGetTenants.png[]
+
+An `Execute` button will appear, and clicking on it sends a GET request that gets all the saved Tenants in JSON format.
+Notice the GET request is sent to the `Request URL`, and the response is found in the `Response body`.
+The server returns the `200` code, which means the request was handled with no issues.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: In my opinion, you are spending too much time explaining how to use the Swagger interface. Can we assume they know? Can we point them to Swagger documentation?

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+
+Nice, we just sent our first HTTP request!
+But the Employee Rostering application does much more than just getting entities.
+There are methods for creating, deleting, and updating various entities (remember the Domain Model above?), which the application uses to modify the database, as well as methods that interact with the OptaPlanner API.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: `Domain Model` should perhaps become a link?

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+Let's try a more interesting method:
+
+image::solveMethod.png[]
+
+This method is the core of the OptaWeb Employee Rostering application.
+It starts the OptaPlanner solver that assigns work shifts to employees for a specific Tenant.
+The solver uses heuristics and algorithms to find the most optimal solution under time and resource constraints.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: Good place to link to Optaplanner documentation for local search, I think.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+The score is calculated using the Drools business constraint rules that we configured in the application.
+As the algorithm finds new solutions, the engine evaluates the score of each solution and compares it with the score of the current best solution.
+If the score is higher, the new solution becomes the new best solution.
+We can see this process in the messages the logger produces during solving:
+
+image::solverLog.png[]
+
+Congratulations!
+We've just found an optimal employee shift arrangement given the time and business constraints for this Tenant.
+Notice how the medium score at the start of solving (-660) was improved to (0) at the end.
```
Comment (triceo) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: You may want to translate this to a language used by humans. What does it mean, for the score to change? Give an example, such as "where we needed 5 employees before, we now only need 3".

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc (hunk 2 above)
Comment (cuijulian) [reply blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: Very good catch! I agree, present perfect tense works better here.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc (hunk 5 above)
Comment (cuijulian) [reply blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: The goal here is not so much to show how to use the Swagger interface, but to show the basics of an HTTP request (what the request does, where the request is sent, what the request returns).

I think Swagger shows these basics quite well, which is why I took the readers through the step-by-step process. As I'm writing this, I realize that showing how an HTTP request works is probably not in the scope of this blog. I'll shorten this so we only show the results of the HTTP request.

Showing Swagger docs isn't the goal, so I'm going to assume readers know what an HTTP request is.

**Comment by cuijulian** [comment]
> Good job, @cuijulian! I am adding a couple comments - please understand that since I'm reviewing a text, the comments are subjective and by no means something definitive that you absolutely need to follow. :-)
> 
> One general comment... You're "selling" the article as a rewrite from Wildfly to Spring Boot. To me, it reads more like an intro to the example app. It certainly doesn't deal with the rewrite itself. Maybe a better name would be `What's new in Employee Rostering?`, `OptaWeb Employee Rostering moves ahead`, or something like that.

Thank you for the review! Your general comment is very insightful, I did struggle to figure out what the main idea of the blog is. I focus on new features of the refactored app rather than the refactoring itself. Since this blog post will likely be a two part series (@Christopher-Chianelli writing about the frontend and I'm writing about the backend), what about `What's New in OptaWeb Employee Rostering? (part 2) - Migrating Backend to Spring Boot`?

**Comment by triceo** [comment]
> Thank you for the review @triceo ! Your general comment is very insightful, I did struggle to figure out what the main idea of the blog is. I focus on new features of the refactored app rather than the refactoring itself. Since this blog post will likely be a two part series (@Christopher-Chianelli writing about the frontend and I'm writing about the backend), what about `What's New in OptaWeb Employee Rostering? (part 2) - Migrating Backend Application to Spring Boot`?

Thinking about it again, I think we could use something a little more buzzword-y. `What's New in OptaWeb Employee Rostering, part 2: Makind the backend cloud-ready`? Perhaps others will have better ideas.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc (hunk 8 above)
Comment (cuijulian) [reply blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: > > Thank you for the review @triceo ! Your general comment is very insightful, I did struggle to figure out what the main idea of the blog is. I focus on new features of the refactored app rather than the refactoring itself. Since this blog post will likely be a two part series (@Christopher-Chianelli writing about the frontend and I'm writing about the backend), what about `What's New in OptaWeb Employee Rostering? (part 2) - Migrating Backend Application to Spring Boot`?
> 
> Thinking about it again, I think we could use something a little more buzzword-y. `What's New in OptaWeb Employee Rostering, part 2: Makind the backend cloud-ready`? Perhaps others will have better ideas.

Haha I'm not the best person for buzzword generation, but your suggestion sure beats mine.

**Comment by cuijulian** [comment]
> > Thank you for the review @triceo ! Your general comment is very insightful, I did struggle to figure out what the main idea of the blog is. I focus on new features of the refactored app rather than the refactoring itself. Since this blog post will likely be a two part series (@Christopher-Chianelli writing about the frontend and I'm writing about the backend), what about `What's New in OptaWeb Employee Rostering? (part 2) - Migrating Backend Application to Spring Boot`?
> 
> Thinking about it again, I think we could use something a little more buzzword-y. `What's New in OptaWeb Employee Rostering, part 2: Makind the backend cloud-ready`? Perhaps others will have better ideas.

Haha I'm not the best person for buzzword generation, but your suggestion sure beats mine. We'll see what other reviewers say.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+:page-interpolate: true
+:awestruct-author: jucui
+:awestruct-layout: blogPostBase
+:awestruct-tags: [employee rostering, coding, algorithm, production]
+
+Recently, the OptaPlanner team has been shifting towards cloud-native applications.
+As a part of this effort, we migrated the OptaWeb Employee Rostering Backend Application from WildFly to Spring Boot.
+
+OptaWeb Employee Rostering is a web application that optimizes shift scheduling for maximum efficiency and fairness for employees.
+In this post we’ll give an overview of the new Spring backend application, and the key features that were added on top of the existing backend.
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: we’ll -> we'll
The ' is different. Use this type of '. Word tend to throw in the other kind.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+== The Problem
+
+Let’s start with the problem we’re trying to solve.
+Here is a visual overview of the https://www.optaplanner.org/learn/useCases/employeeRostering.html[Employee Rostering problem]:
+
+image::employeeRosteringProblem.png[]
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: It's very good to use many images.
It's good to reuse images from the docs.
It's best to indeed to copy them here, practically speaking.

Can you keep the same file name though if they are unaltered? Do give them a different name if they altered.
If we update those employeeRosteringValueProposition.png in the docs, it might worth updating this blog post too, in a few years down the line. (Blog post can live very long).

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+== Domain Model
+
+Before we dive into the code, we first need a domain model.
+A domain model represents the entities (Java objects) and how they relate with other entities in the application.
+It tells us the behaviour of the objects and how they change during solving.
+Here's a class diagram of the Employee Rostering model, where employees are nurses:
+
+image::employeeRosteringClassDiagram.png[]
+
+Each of the blocks above represents a Java entity object in our application, which will be saved in a database for solving.
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: which will be saved -> are saved
Prefer present tense over future tense whenever possible.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+----
+
+This Java object stores the name of the skill and the id of the Tenant (the user, such as a hospital or a call center) it belongs to.
+The `@Entity` annotation represents a table in the JPA database, where each instance of the `Skill` entity has its own row.
+The `@Table` annotation specifies the columns of the table (we'll see this in the next section).
+
+== H2 Database
+
+To test the database, we use the H2 Database Console to run SQL statements and view the saved entities on a user interface.
+Here's an example of executing a SQL statement that will select every `Skill` entity:
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: will select -> selects
see https://developers.google.com/style/tense

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+image::skillsConsole.png[]
+
+Here we see the autogenerated demo data saved in the H2 database when the application starts up.
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: Don't use "we", nor "I". Use "you" if you have to. Best is to use active tense whenever possible and make it shorter.
"Here we see" -> "Here you see ..." or "Notice that ... is saved"

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc (hunk 7 above)
Comment (ge0ffrey) [reply blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: Linking into sections of the optaplanner docs is maintenance unfriendly, unfortunately. (It shouldn't be, but it is what it is.)
You can use the "latest" link to avoid hard coding a version number, for example https://docs.optaplanner.org/latestFinal/optaplanner-docs/html_single/  but I generally avoid linking into the user guide for blogs.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+
+Let's try out the solver for the Tenant with `id = 1`:
+
+image::executeSolverMethod.png[]
+
+After clicking `Execute`, the solver starts evaluating possible solutions for arranging shifts, and finds the shift roster with the best score.
+The score is calculated using the Drools business constraint rules that we configured in the application.
+As the algorithm finds new solutions, the engine evaluates the score of each solution and compares it with the score of the current best solution.
+If the score is higher, the new solution becomes the new best solution.
+We can see this process in the messages the logger produces during solving:
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: Try to remove all the sentence parts that are just ceremony. Add where those logging message appear (= in the console).
-> "Notice the log messages during solving in the console:"

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+
+image::solverLog.png[]
+
+Notice the medium score when solving starts versus when solving ends.
+In our application, each unassigned shift has a score of (-1medium), which means for every unassigned shift a particular solution has, the medium score of that solution would decrease by 1.
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: use ` around -1medium

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+In this case, the starting shift arrangement had a medium score of -660, so there were 660 shifts that were not assigned to any employee.
+After solving, the shift arrangement had a medium score of 0, which means every shift was assigned to an employee.
+
+Congratulations!
```
Comment (ge0ffrey) [inline blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: You're writing in the we-form, so _you've just congratulated yourself_ :)

It's ok to keep that congratulations, once it's transformed into the you-form.

File: blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc (hunk 9 above)
Comment (cuijulian) [reply blog/2019-10-02-MigratingOptaWebEmployeeRosteringToSpringBoot.adoc]: Strange, I have the correct "we'll" in my IDE, but when I push GitHub shows the former symbol. Is there an IDE setting I'm missing?

**Comment by cuijulian** [comment]
> This is a really nice blog! Great work Julian!
> 
> If you can process the comments, adjust the filename for the new date (I suggest Tuesday), I 'll merge it and it will automatically publish.

Thanks Geoffrey! I've made changes according to your comments (new date is next Tues, October 22). It's good to merge for Tuesday.
